tests/element_test_vectors.rs

```diff
@@ -34,60 +37,126 @@ fn contains_path(paths: &[&str], file_name: &str) -> bool {
+//       native writer so we can fix any failures it produces.
+impl RoundTrip for NativeElementWriterApi {
+    fn roundtrip<R: ElementReader>(
+        elements: &Vec<OwnedElement>,
+        format: Format,
+        reader: R,
+    ) -> IonResult<Vec<OwnedElement>> {
+        // Unlike the C writer, the Rust writer can write into a growing Vec.
+        // No need for an aggressive preallocation.
+        let mut buffer = Vec::with_capacity(2048);
+        match format {
+            // TODO: Handle `TextKind`: Pretty
+            Format::Text(_) => {
+                let raw_text_writer = RawTextWriter::new(&mut buffer);
+                let mut writer = NativeElementWriter::new(TextWriter::new(raw_text_writer));
+                writer.write_all(elements)?;
+                let _ = writer.finish()?;
+            }
+            Format::Binary => {
+                let raw_binary_writer = RawBinaryWriter::new(&mut buffer);
+                let mut writer = NativeElementWriter::new(BinaryWriter::new(raw_binary_writer));
+                writer.write_all(elements)?;
+                let _ = writer.finish()?;
+            }
+        };
+
+        reader.read_all(buffer.as_slice())
+    }
+}
+
 trait ElementApi {
-    type ElementReader: ElementReader;
+    type ReaderApi: ElementReader;
+    type RoundTripper: RoundTrip;
 
     fn global_skip_list() -> SkipList;
     fn read_one_equivs_skip_list() -> SkipList;
     fn round_trip_skip_list() -> SkipList;
     fn equivs_skip_list() -> SkipList;
     fn non_equivs_skip_list() -> SkipList;
 
-    fn make_reader() -> Self::ElementReader;
-
-    // TODO: These are currently all ion-c specific method signatures. SliceElementWriter takes a
-    //       lifetime, so we can't specify it as an associated type until GATs is stable. That's
-    //       ok for the moment because we don't have a native ElementWriter to test yet.
-    fn make_text_writer(buffer: &mut [u8]) -> IonResult<SliceElementWriter>;
-    fn make_pretty_text_writer(buffer: &mut [u8]) -> IonResult<SliceElementWriter>;
-    fn make_binary_writer(buffer: &mut [u8]) -> IonResult<SliceElementWriter>;
+    fn make_reader() -> Self::ReaderApi;
 
     /// Asserts the given elements can be round-tripped and equivalent, then returns the new elements.
-    fn assert_round_trip<F>(
+    fn assert_round_trip(
         source_elements: &Vec<OwnedElement>,
-        make_writer: F,
-    ) -> IonResult<Vec<OwnedElement>>
-    where
-        F: FnOnce(&mut [u8]) -> IonResult<SliceElementWriter>,
-    {
-        let mut buf = vec![0u8; WRITE_BUF_LENGTH];
-        let mut writer = make_writer(&mut buf)?;
-        writer.write_all(source_elements)?;
-        let output = writer.finish()?;
-        let new_elements = element_reader().read_all(output)?;
+        format: Format,
+    ) -> IonResult<Vec<OwnedElement>> {
+        let new_elements =
+            Self::RoundTripper::roundtrip(source_elements, format, Self::make_reader())?;
```

Comment:
🗺️ Blink and you'll miss it! I know because _I_ missed it. 😞 

The `Self::make_reader()` at the end of this line used to be a call to `element_reader()`, which meant that for lots of tests the native Rust reader wasn't being used; `element_reader()` always returns an `IonCSliceElementWriter`.

This PR fixed that, bringing a handful of other test failures to light. This PR also fixed most of them.